internal/gapicgen/generator/generator.go

```diff
@@ -27,41 +27,64 @@ import (
 )
 
 // Generate generates genproto and gapics.
-func Generate(ctx context.Context, googleapisDir, genprotoDir, gocloudDir, protoDir string, gapicToGenerate string) error {
-	if err := regenGenproto(ctx, genprotoDir, googleapisDir, protoDir); err != nil {
-		return fmt.Errorf("error generating genproto (may need to check logs for more errors): %v", err)
+func Generate(ctx context.Context, googleapisDir, genprotoDir, gocloudDir, protoDir string, gapicToGenerate string) ([]*ChangeInfo, error) {
+	protoGenerator := NewGenprotoGenerator(genprotoDir, googleapisDir, protoDir)
+	gapicGenerator := NewGapicGenerator(googleapisDir, protoDir, gocloudDir, genprotoDir)
+	if err := protoGenerator.Regen(ctx); err != nil {
+		return nil, fmt.Errorf("error generating genproto (may need to check logs for more errors): %v", err)
+	}
+	if err := gapicGenerator.Regen(ctx, gapicToGenerate); err != nil {
+		return nil, fmt.Errorf("error generating gapics (may need to check logs for more errors): %v", err)
 	}
 
-	if err := generateGapics(ctx, googleapisDir, protoDir, gocloudDir, genprotoDir, gapicToGenerate); err != nil {
-		return fmt.Errorf("error generating gapics (may need to check logs for more errors): %v", err)
+	changes, err := gatherChanges(googleapisDir, genprotoDir)
+	if err != nil {
+		return nil, fmt.Errorf("error gathering commit info")
 	}
 
 	if err := recordGoogleapisHash(googleapisDir, genprotoDir); err != nil {
-		return fmt.Errorf("error recording most recent googleapis hash: %v", err)
+		return nil, err
 	}
 
-	return nil
+	return changes, nil
+}
+
+func gatherChanges(googleapisDir, genprotoDir string) ([]*ChangeInfo, error) {
+	// Get the last processed googleapis hash.
+	lastHash, err := ioutil.ReadFile(filepath.Join(genprotoDir, "regen.txt"))
+	if err != nil {
+		return nil, err
+	}
+	commits, err := CommitsSinceHash(googleapisDir, string(lastHash), true)
+	if err != nil {
+		return nil, err
+	}
+	var changes []*ChangeInfo
+	for _, v := range commits {
+		c, err := ParseChangeInfo(googleapisDir, v)
+		if err != nil {
+			return nil, err
+		}
+		changes = append(changes, c)
+	}
+
+	return changes, nil
 }
 
 // recordGoogleapisHash parses the latest commit in googleapis and records it to
 // regen.txt in go-genproto.
 func recordGoogleapisHash(googleapisDir, genprotoDir string) error {
-	out := bytes.NewBuffer(nil)
-	c := command("git", "rev-list", "HEAD^..")
-	c.Stdout = out
-	c.Stderr = os.Stderr
-	c.Dir = googleapisDir
-	if err := c.Run(); err != nil {
+	commits, err := CommitsSinceHash(googleapisDir, "HEAD", true)
+	if err != nil {
 		return err
 	}
-	commits := strings.Split(strings.TrimSpace(out.String()), "\n")
 	if len(commits) != 1 {
 		return fmt.Errorf("only expected one commit, got %d", len(commits))
 	}
 
 	f, err := os.OpenFile(filepath.Join(genprotoDir, "regen.txt"), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
 	if err != nil {
-		return err
+		log.Fatal(err)
```

Comment:
Why `log.Fatal`?

Reply:
Left in from when I was testing something, good catch!